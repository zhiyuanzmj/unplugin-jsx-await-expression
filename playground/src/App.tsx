import { Suspense, defineComponent, ref } from 'vue'

const getList = async (length: number) => {
  await new Promise((r) => setTimeout(r, 1000))
  if (length > 4) throw new Error(`can't large than 3`)
  return Array.from({ length }).map((_, i) => i + 1)
}

const Comp = defineComponent(async () => {
  const length = ref(3)
  const foo = ref(0)

  return (
    <>
      {length.value}
      <button onClick={() => length.value++}>+</button>
      <button onClick={() => length.value--}>-</button>
      <Suspense v-slots={{ fallback: () => <div>Loading...</div> }}>
        {await getList(length.value)
          .then((items) => {
            return items.map((i) => (
              <div onClick={() => foo.value++}>
                {i}
                {foo.value}
              </div>
            ))
          })
          .catch((error) => <>{error.message}</>)}
      </Suspense>
    </>
  )
})

export default defineComponent(() => {
  return (
    <Suspense>
      <Comp />
    </Suspense>
  )
})
