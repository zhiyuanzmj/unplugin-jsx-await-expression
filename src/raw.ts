import {
  HELPER_PREFIX,
  MagicStringAST,
  REGEX_SETUP_SFC,
  REGEX_SRC_FILE,
  createFilter,
  generateTransform,
  getLang,
  importHelperFn,
  parseSFC,
  walkAST,
} from '@vue-macros/common'
import { type Options, resolveOption } from './core/options'
import type { Node } from 'oxc-parser'
import type { UnpluginOptions } from 'unplugin'

const { parseSync } = await import(
  // @ts-ignore
  typeof window !== 'undefined'
    ? 'https://cdn.jsdelivr.net/npm/@oxc-parser/binding-wasm32-wasi/browser-bundle.mjs'
    : 'oxc-parser'
)

function transform(code: string, s: MagicStringAST) {
  const { program } = parseSync('index.tsx', code, {
    sourceType: 'module',
  })

  walkAST<Node>(program, {
    enter(node) {
      if (
        node.type === 'JSXExpressionContainer' &&
        node.expression.type === 'AwaitExpression'
      ) {
        const { expression } = node
        const createVNode = importHelperFn(s, 0, 'createVNode')
        const shallowRef = importHelperFn(s, 0, 'shallowRef')
        const watchEffect = importHelperFn(s, 0, 'watchEffect')
        let resolvedFn = ''
        let injectedFn = ''
        s.appendRight(
          expression.start,
          `${createVNode}({async setup() {
            const ${HELPER_PREFIX}resolved = ${shallowRef}()
            const ${HELPER_PREFIX}isInject = ${shallowRef}()
            await new Promise((resolve) =>
              ${watchEffect}(async () => resolve(`,
        )
        const { argument } = expression
        processArgument(argument)
        function processArgument(argument: Node) {
          if (
            argument.type === 'CallExpression' &&
            argument.callee.type === 'MemberExpression'
          ) {
            const {
              callee: { property, object },
              arguments: [arg],
            } = argument
            if (
              property.type === 'Identifier' &&
              (arg?.type === 'ArrowFunctionExpression' ||
                arg?.type === 'FunctionExpression')
            ) {
              if (property.name === 'then' && !resolvedFn) {
                resolvedFn = `(${s.sliceNode(arg)})`
                s.overwriteNode(
                  arg.body!,
                  `{${HELPER_PREFIX}isInject.value = false; ${HELPER_PREFIX}resolved.value = ${s.sliceNode(arg.params[0])} }`,
                )
              } else if (property.name === 'catch' && !injectedFn) {
                injectedFn = `(${s.sliceNode(arg)})`
                s.overwriteNode(
                  arg.body!,
                  `{${HELPER_PREFIX}isInject.value = true; ${HELPER_PREFIX}resolved.value = ${s.sliceNode(arg.params[0])} }`,
                )
              }
            }
            if (object.type === 'CallExpression') {
              processArgument(object)
            }
          }
        }

        if (!injectedFn && !resolvedFn) {
          s.appendLeft(
            expression.end,
            `.then(i => ${HELPER_PREFIX}resolved.value = i)`,
          )
        }
        s.appendLeft(
          expression.end,
          `)));return () => ${injectedFn ? `${HELPER_PREFIX}isInject.value ? (${injectedFn})(${HELPER_PREFIX}resolved.value) :` : ''} ${resolvedFn}(${HELPER_PREFIX}resolved.value)`,
        )
        s.appendLeft(expression.end, `}})`)
      }
    },
  })
}

export function transformJSXAwaitExpression(code: string, id: string) {
  const lang = getLang(id)
  let asts: {
    text: string
    offset: number
  }[] = []
  if (lang === 'vue' || REGEX_SETUP_SFC.test(id)) {
    const { scriptSetup, script } = parseSFC(code, id)
    if (script?.content) {
      asts.push({
        text: script.content,
        offset: script.loc.start.offset,
      })
    }
    if (scriptSetup) {
      asts.push({
        text: scriptSetup.content!,
        offset: scriptSetup.loc.start.offset,
      })
    }
  } else if (REGEX_SRC_FILE.test(id)) {
    asts = [{ text: code, offset: 0 }]
  } else {
    return
  }

  const s = new MagicStringAST(code)

  for (const { text, offset } of asts) {
    s.offset = offset
    transform(text, s)
  }

  return generateTransform(s, id)
}

const plugin = (rawOptions: Options = {}): UnpluginOptions => {
  const options = resolveOption(rawOptions)
  const filter = createFilter(options)

  const name = 'unplugin-jsx-await-expression'
  return {
    name,
    enforce: 'pre',

    transformInclude(id) {
      return filter(id)
    },
    transform: transformJSXAwaitExpression,
  }
}
export default plugin
