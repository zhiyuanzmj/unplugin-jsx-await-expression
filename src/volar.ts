import { createPlugin, replaceSourceRange } from 'ts-macro'

export default createPlugin(({ ts }) => {
  return {
    name: 'jsx-await-expression',
    resolveVirtualCode({ codes, ast, source }) {
      ts.forEachChild(ast, walk)

      function walk(node: import('typescript').Node) {
        ts.forEachChild(node, walk)
        if (
          ts.isJsxExpression(node) &&
          node.expression &&
          ts.isAwaitExpression(node.expression)
        ) {
          replaceSourceRange(
            codes,
            source,
            node.expression.pos,
            node.expression.expression.pos,
          )
        }
      }
    },
  }
})
