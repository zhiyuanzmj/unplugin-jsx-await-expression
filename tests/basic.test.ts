import { describe, expect, test } from 'vitest'
import { transformJSXAwaitExpression } from '../src/raw'

describe('fixtures', async () => {
  test('basic', () => {
    const { code } = transformJSXAwaitExpression(
      `
defineComponent(()=>{
  const getList = async () => {
    await new Promise((r) => setTimeout(r, 1000))
    return [1, 2, 3]
  }

  return () => <>{await getList(length.value)
          .then(items => items)
          .then((items) => items.map((i) => <div>{i}{length.value}</div>))
          .catch((e) => <>{e.message}</>)
          }</>
})
      `,
      'index.tsx',
    )
    expect(code).toMatchInlineSnapshot(`
      "
      import { createVNode as __MACROS_createVNode } from "vue";
      import { shallowRef as __MACROS_shallowRef } from "vue";
      import { watchEffect as __MACROS_watchEffect } from "vue";
      defineComponent(()=>{
        const getList = async () => {
          await new Promise((r) => setTimeout(r, 1000))
          return [1, 2, 3]
        }

        return () => <>{__MACROS_createVNode({async setup() {
                  const __MACROS_resolved = __MACROS_shallowRef()
                  const __MACROS_isInject = __MACROS_shallowRef()
                  await new Promise((resolve) =>
                    __MACROS_watchEffect(async () => resolve(await getList(length.value)
                .then(items => items)
                .then((items) => {__MACROS_isInject.value = false; __MACROS_resolved.value = items })
                .catch((e) => {__MACROS_isInject.value = true; __MACROS_resolved.value = e }))));return () => __MACROS_isInject.value ? (((e) => <>{e.message}</>))(__MACROS_resolved.value) : ((items) => items.map((i) => <div>{i}{length.value}</div>))(__MACROS_resolved.value)}})
                }</>
      })
            "
    `)
  })
})
