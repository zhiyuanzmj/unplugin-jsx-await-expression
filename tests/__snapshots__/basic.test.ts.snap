// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > tests/fixtures/basic.tsx 1`] = `
"import { defineComponent, createVNode, Fragment } from 'vue';

var basic = /* @__PURE__ */ defineComponent(() => {
  const getList = async () => {
    await new Promise(r => setTimeout(r, 1000));
    return [1, 2, 3];
  };
  return () => createVNode(Fragment, null, [getList()]);
});

export { basic as default };
"
`;
